api_core/google/api_core/protobuf_helpers.py

```diff
@@ -76,3 +82,150 @@ def get_messages(module):
+
+    # If the object we got back is our sentinel, raise KeyError; this is
+    # a "not found" case.
+    if answer is _SENTINEL:
+        raise KeyError(key)
+
+    # If a subkey exists, call this method recursively against the answer.
+    if subkey and answer is not default:
+        return get(answer, subkey, default=default)
+
+    return answer
+
+
+def set(msg_or_dict, key, value):
+    """Set a key's value on a protobuf Message or dictionary.
+
+    Args:
+        msg_or_dict (Union[~google.protobuf.message.Message, Mapping]): the
+            object.
+        key (str): The key to set.
+        value (Any): The value to set.
+
+    Raises:
+        TypeError: If ``msg_or_dict`` is not a Message or dictionary.
+    """
+    # Sanity check: Is our target object valid?
+    if not isinstance(msg_or_dict, (collections.MutableMapping, Message)):
+        raise TypeError(
+            'set() expected a dict or protobuf message, got {!r}.'.format(
+                type(msg_or_dict)))
+
+    # We may be setting a nested key. Resolve this.
+    key, subkey = _resolve_subkeys(key)
+
+    # If a subkey exists, then get that object and call this method
+    # recursively against it using the subkey.
+    if subkey is not None:
+        if isinstance(msg_or_dict, collections.MutableMapping):
+            msg_or_dict.setdefault(key, {})
+        set(get(msg_or_dict, key), subkey, value)
+        return
+
+    # Attempt to set the value on the types of objects we know how to deal
+    # with.
+    if isinstance(msg_or_dict, collections.MutableMapping):
+        msg_or_dict[key] = value
+    elif isinstance(value, (collections.MutableSequence, tuple)):
+        # Clear the existing repeated protobuf message of any elements
+        # currently inside it.
+        while getattr(msg_or_dict, key):
+            getattr(msg_or_dict, key).pop()
+
+        # Write our new elements to the repeated field.
+        for item in value:
+            if isinstance(item, collections.Mapping):
+                getattr(msg_or_dict, key).add(**item)
+            else:
+                getattr(msg_or_dict, key).extend([item])
+    elif isinstance(value, collections.Mapping):
+        # Assign the dictionary values to the protobuf message.
+        for item_key, item_value in value.items():
+            set(getattr(msg_or_dict, key), item_key, item_value)
+    elif isinstance(value, Message):
+        # Assign the protobuf message values to the protobuf message.
+        for item_key, item_value in value.ListFields():
+            set(getattr(msg_or_dict, key), item_key.name, item_value)
+    else:
+        setattr(msg_or_dict, key, value)
+
+
+def setdefault(msg_or_dict, key, value):
+    """Set the key on a protobuf Message or dictioanary to a given value if the
+    current value is falsy.
+
+    Because protobuf Messages do not distinguish between unset values and
+    falsy ones particularly well, this method treats any falsy value
```

Comment:
"particularly well (by design), this ..."

Reply:
Done.